libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,722 @@
+	return exp[n];
+}
+
+inline uint8_t nElem(uint8_t _b)
+{
+	return pow2N((_b) & 0xf);
+}
+
+inline uint8_t elemType(uint8_t _b)
+{
+	return (_b) >> 4;
+}
+
+u256 VM::vtow(uint8_t _b, const u256& _in)
+{
+	u256 out;
+	uint8_t const n = nElem(_b);
+	uint8_t const t = elemType(_b);
+	switch (t)
+	{
+	case 0:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 8;
+			out |= v8x32(_in) [i];
+		}
+		break;
+	case 1:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 16;
+			out |= v16x16(_in)[i];
+		}
+		break;
+	case 2:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 32;
+			out |= v32x8(_in) [i];
+		}
+		break;
+	case 3:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 64;
+			out |= v64x4(_in) [i];
+		}
+		break;
+	default:
+		throwBadInstruction();
+	}
+	return out;
+}
+
+u256 VM::wtov(uint8_t _b, const u256& _in)
+{
+	u256 out;
+	uint8_t const n = nElem(_b);
+	uint8_t const t = elemType(_b);
+	switch (t)
+	{
+	case 0:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v8x32(out) [i] |= (uint8_t )(_in & 0xff);
+			_in >>= 8;
+		}
+		break;
+	case 1:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v16x16(out)[i] |= (uint16_t)(_in & 0xffff);
+			_in >>= 16;
+		}
+		break;
+	case 2:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v32x8(out) [i] |= (uint32_t)(_in & 0xffffff);
+			_in >>= 32;
+		}
+		break;
+	case 3:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v64x4(out) [i] |= (uint64_t)(_in & 0xffffffff);
+			_in >>= 64;
+		}
+		break;
+	default:
+		throwBadInstruction();
+	}
+	return out;
+}
+
+void VM::xmload (uint8_t _b)
+{
+	// n bytes of type t elements in memory vector
+	// goes onto stack element by element, LSB first
+	uint8_t* p = m_mem.data() + toInt15(m_SP[0]);
+	uint8_t const n = nElem(_b);
+	uint8_t const t = elemType(_b);
+
+	switch (t)
+	{
+	case 0:
+		for (int j = n, v = 0, i = 0; 0 < n; ++i)
+		{
+			v |= p[--j];
+			v8x32(m_SPP[0])[i] = v;
+		}
+		break;
+	case 1:
+		for (int j = n, v = 0, i = 0; 0 < n; ++i)
+		{
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v16x16(m_SPP[0])[i] = v;
+		}
+		break;
+	case 2:
+		for (int j = n,  v = 0, i = n - 1; 0 <= i; --i)
+		{
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v32x8(m_SPP[0])[i] = v;
+		}
+		break;
+	case 3:
+		for (int j = n,  v = 0, i = n - 1; 0 <= i; --i)
+		{
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v <<= 8;
+			v |= p[--j];
+			v64x4(m_SPP[0])[i] = v;
+		}
+		break;
+	default:
+		throwBadInstruction();
+	}
+}
+   
+void VM::xmstore(uint8_t _b)
+{
+	// n bytes of type t elements in stack vector
+	// goes onto memory by element, LSB first
+	uint8_t *p = m_mem.data() + toInt15(m_SP[0]);
+	uint8_t const n = nElem(_b);
+	uint8_t const t = elemType(_b);
+
+	switch (t)
+	{
+	case 0:
+		for (int j = n, v = 0, i = 0; 0 < n; ++i)
```

Comment:
Yeah, I make that mistake in four places - it should be 'i < n'.  But why am I counting up there but down after?   I realized I couldn't remember how these worked - so set them aside as I fixed easier things. Now I have this problem preventing compilation with the EIP's off - but only on Windows.